src/semantics.js

```diff
@@ -160,73 +173,107 @@ var semicolonCmdNames = [
-        ') {' + nl(this.args.indent+1) + cmd.toJS(this.args.indent+1);
+    return nl(this.args.indent) + '} else if (' + cond.toJS(this.args.indent, this.args.ctx) +
+        ') {' + nl(this.args.indent+1) + cmd.toJS(this.args.indent+1, this.args.ctx);
   },
   ElseCase: function(_sc, _else, _space, cmd) {
     return nl(this.args.indent) + '} else {' + nl(this.args.indent+1) +
-        cmd.toJS(this.args.indent+1);
+        cmd.toJS(this.args.indent+1, this.args.ctx);
   },
   EndIf: function(_sc, _fi) {
     return nl(this.args.indent) + '}';
   },
-  ForCommand: function(f) { return f.toJS(this.args.indent); },
+  ForCommand: function(f) { return f.toJS(this.args.indent, this.args.ctx); },
   ForCommand_c_style: function(_for, _op, ctrlstruct, _cp, _sc3, _do, _s, cmd, done) {
-    return 'for (' + ctrlstruct.toJS(0) + ') {' +
-        nl(this.args.indent+1) + cmd.toJS(this.args.indent+1) +
+    return 'for (' + ctrlstruct.toJS(0, this.args.ctx) + ') {' +
+        nl(this.args.indent+1) + cmd.toJS(this.args.indent+1, this.args.ctx) +
         nl(this.args.indent) + '}';
   },
   ControlStruct: function(assign, _sc1, id, binop, val, _sc2, update) {
-    return assign.toJS(0) + ';' + id.sourceString + binop.toJS(0) + val.toJS(0) +
+    return assign.toJS(0, this.args.ctx) + ';' + id.sourceString + binop.toJS(0, this.args.ctx) + val.toJS(0, this.args.ctx) +
       ';' + update.sourceString;
   },
   ForCommand_for_each: function(_for, id, _in, call, _sc, _do, _s, cmd2, done) {
-    var mycall = call.toJS(this.args.indent).replace(/\.replace\(.*\)/, '');
+    var mycall = call.toJS(this.args.indent, this.args.ctx).replace(/\.replace\(.*\)/, '');
     return mycall + '.forEach(function (' + id.sourceString + ') {' +
-        nl(this.args.indent+1) + cmd2.toJS(this.args.indent+1) +
+        nl(this.args.indent+1) + cmd2.toJS(this.args.indent+1, this.args.ctx) +
         nl(this.args.indent) + '});';
   },
   WhileCommand: function(_w, _s, cond, _sc, _do, _s2, cmd, done) {
-    return 'while (' + cond.toJS(this.args.indent) + ') {' +
-        nl(this.args.indent+1) + cmd.toJS(this.args.indent+1) +
-        done.toJS(this.args.indent);
+    return 'while (' + cond.toJS(this.args.indent, this.args.ctx) + ') {' +
+        nl(this.args.indent+1) + cmd.toJS(this.args.indent+1, this.args.ctx) +
+        done.toJS(this.args.indent, this.args.ctx);
   },
   Done: function(_sc, _) {
     return nl(this.args.indent) + '}';
   },
   FunctionDecl: function(_fun, _sp1, id, _paren, _sp2, block) {
-    var idStr = id.toJS(0);
+    var idStr = id.toJS(0, this.args.ctx);
     allFunctions[idStr] = true;
 
     inFunctionBody = true;
-    var blockString = block.toJS(this.args.indent);
+    var blockString = block.toJS(this.args.indent, this.args.ctx);
     inFunctionBody = false;
 
     return 'function ' + idStr + '(..._$args) ' + blockString;
   },
+  CaseCommand: function(_case, expr, _in, _ws, cases, _ws2, _esac){
+    var varName = "tmp" + Date.now();
```

Comment:
Would it be possible to change this to something more consistent? It makes testing a lot easier. If you do something like prefix some name with a `$`, then it should be safe to use in the translation (`$` can't be in a bash identifier, but it works for JS).